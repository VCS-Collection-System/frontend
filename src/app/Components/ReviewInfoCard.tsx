import React, { useState } from 'react';
import { 
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionToggle,
  Card, 
  CardBody, 
  Form, 
  FormGroup, 
  TextInput 
} from '@patternfly/react-core';
import CardWithTitle from './CardWithTitle';
import { dateConvert } from "@app/utils/utils";

const ReviewInfoCard = (props) => {

  const CovidTestInfo = () => {
    return (
      <>
        <FormGroup
          label="Test Result or Submission Type"
          fieldId="test-result">
          <TextInput
            id="test-result"
            isDisabled
            type="text"
            value={props.covidTestResult}
          />
        </FormGroup> 
        <FormGroup
          label="Effective Date"
          fieldId="test-date">
          <TextInput
            id="test-date"
            isDisabled
            type="text"
            value={dateConvert(props.covidTestDate)}
          />
        </FormGroup>
      </>
    )
  }

  const CovidVaccineInfo = () => {
    const [expanded, setExpanded] = useState(["shot-" + (props.doses.length-1) + "-toggle"]);

    const toggle = (id) => {
      const index = expanded.indexOf(id);
      const newExpanded: string[] =
        index >= 0 ? [...expanded.slice(0, index), ...expanded.slice(index + 1, expanded.length)] : [...expanded, id];
      setExpanded(newExpanded);
    };

    return (
      <Accordion asDefinitionList={false} className="pf-m-bordered">
        {props.doses.map((dose, i) => (
          <AccordionItem key={i}>
            <AccordionToggle
              onClick={() => toggle("shot-" + i + "-toggle")}
              isExpanded={expanded.includes("shot-" + i + "-toggle")}
              id={"shot-" + i + "-toggle"}
            >
              Shot {i+1}
            </AccordionToggle>
            <AccordionContent id={"shot-" + i + "-content"} isHidden={!expanded.includes("shot-" + i + "-toggle")}>
              <FormGroup
                label="Vaccine"
                fieldId={"vaccine-brand-" + i}
              >
                <TextInput
                  id={"vaccine-brand-" + i}
                  isDisabled
                  type="text"
                  value={dose.vaccineBrand}
                />
              </FormGroup>
              <FormGroup
                label="Date Administered"
                fieldId={"vaccine-date-administered-" + i}
                style={{marginTop: "1rem"}}
              >
                <TextInput
                  id={"vaccine-date-administered-" + i}
                  isDisabled
                  type="text"
                  value={dateConvert(dose.vaccineDate)}
                />
              </FormGroup>
              <FormGroup
                label="Lot/Batch Number"
                fieldId={"vaccine-lot-number-" + i}
                style={{marginTop: "1rem"}}
              >
                <TextInput
                  id={"vaccine-lot-number-" + i}
                  isDisabled
                  type="text"
                  value={dose.vaccineLotNumber}
                />
              </FormGroup>
            </AccordionContent>
          </AccordionItem>
        ))}
      </Accordion>
    )
  }

  const ReviewTypeInfo = () => {
    if(props.reviewType === "TEST") {
      return <CovidTestInfo />
    }
    else if(props.reviewType === "VAX") {
      return <CovidVaccineInfo />
    }
    else {
      return <></>
    }
  }

  const HRActionInfo = () => {
    if(props.submittedBy) {
      return (
        <FormGroup
          className="warningText"
          label="Submitted By"
          fieldId="profile-form-submitted-by"
        >
          <TextInput
            id="profile-input-submitted-by"
            isDisabled
            type="text"
            value={props.submittedBy}
          />
        </FormGroup>
      )
    }
    else if(props.reviewer) {
      return (
        <FormGroup
          className="warningText"
          label="Reviewed By"
          fieldId="profile-form-submitted-by"
        >
          <TextInput
            id="profile-input-submitted-by"
            isDisabled
            type="text"
            value={props.reviewer}
          />
        </FormGroup>
      )
    }
    else {
      return <></>
    }
  }

  return (
    <Card>
      <CardWithTitle 
        title="Review Information"
      />
      <CardBody>
        <Form>
          <HRActionInfo />
          <ReviewTypeInfo />
        </Form>
      </CardBody>
    </Card>
  )
}

export default ReviewInfoCard;