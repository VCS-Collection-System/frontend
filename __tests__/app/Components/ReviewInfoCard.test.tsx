import React from 'react';
import { render, screen } from '@testing-library/react'
import ReviewInfoCard from '../../../src/app/Components/ReviewInfoCard';

describe('<ReviewInfoCard />', ()=> {
    it('should be defined', () => {
        expect(ReviewInfoCard).toBeDefined()
    })

    it("should match existing snapshot", () => {
        expect(render(<ReviewInfoCard />).asFragment()).toMatchSnapshot()
    })

    it("should display covid test info", () => {
        render(
            <ReviewInfoCard 
                reviewType={"TEST"}
                covidTestResult={"POSITIVE"}
                covidTestDate={"2021-12-10"}
            />
        )

        expect(screen.getByDisplayValue('POSITIVE')).toBeTruthy()
        expect(screen.getByDisplayValue('12-10-2021')).toBeTruthy()
    })

    it("should display covid vax info", () => {
        const doses = [{
            vaccineBrand: "Moderna",
            vaccineDate: "2021-12-10",
            vaccineLotNumber: "lot1"
        }]

        render(
            <ReviewInfoCard 
                reviewType={"VAX"}
                doses={doses}
            />
        )

        expect(screen.getByDisplayValue('Moderna')).toBeTruthy()
        expect(screen.getByDisplayValue('12-10-2021')).toBeTruthy()
        expect(screen.getByDisplayValue('lot1')).toBeTruthy()
    })

    it("should display submit by info", () => {
        render(
            <ReviewInfoCard 
                submittedBy={"Mock Submitter"}
            />
        )

        expect(screen.getByDisplayValue('Mock Submitter')).toBeTruthy()
    })

    it("should display reviewed by info", () => {
        render(
            <ReviewInfoCard 
                reviewer={"Mock Reviewer"}
            />
        )

        expect(screen.getByDisplayValue('Mock Reviewer')).toBeTruthy()
    })
})
