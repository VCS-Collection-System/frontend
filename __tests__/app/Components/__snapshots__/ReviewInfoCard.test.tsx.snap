// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewInfoCard /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Review Information
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        />
      </div>
    </div>
    <div
      class="pf-c-card__body"
    >
      <form
        class="pf-c-form"
        novalidate=""
      />
    </div>
  </article>
</DocumentFragment>
`;
